Add Haskell support (#5281)

pseudomata opened the pull request:
This PR adds the Haskell tree-sitter grammar copied from [`nvim-treesitter`](https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries/haskell). It also adds the Haskell Language Server.

This is a joint effort by myself (adding the grammar) and @leifu1128 (who is adding the language server integration).

This PR resolves https://github.com/zed-industries/zed/issues/5281

Release Notes:

- Added Haskell support ([#5281](https://github.com/zed-industries/zed/issues/5281)).

Commit:
Get basic syntax highlighting for Haskell working

Commit:
Update Haskell grammar

Copied queries from nvim-tree-sitter, see https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries/haskell

Commit:
Merge branch 'haskell-support' into add-haskell-grammar

Commit:
Merge pull request #1 from pseudomata/add-haskell-grammar

Add haskell grammar

leizaf committed:
implement hls adapter

Commit:
Merge pull request #2 from leifu1128/hls-adapter

implement hls adapter

leifu1128 commented:
@cla-bot check

pseudomata commented:
Sweet, thanks @leifu1128! Marking this "Ready for review". There was an open question about which directory the LSP runs in (https://github.com/zed-industries/zed/issues/5281#issuecomment-1912226039) though I do not think that's a blocker and can likely be a follow-up (I don't have a strong preference).

pseudomata commented on crates/zed/src/languages/haskell/highlights.scm:
```diff
@@ -0,0 +1,632 @@
+
+(con_unit) @string.special.symbol ; unit, as in ()
+
+(comment) @comment
```
The `highlights.scm` file had another bit for comments that I removed because it made the comments not render correctly, I am not sure what the purpose of that was but here's a link to the removed lines anyway:

https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/haskell/highlights.scm#L621-L623

mikayla-maki replied:
Thanks for the link! I think in general we'd prefer to remove commented out highlights rather than keep them, so this was great :)

Commit:
Comment out haskell_persistent and sql injections

- the haskell_persistent grammar has not been added, see https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries/haskell_persistent and https://github.com/MercuryTechnologies/tree-sitter-haskell-persistent
- the sql grammar has not been added either

mikayla-maki commented on crates/zed/src/languages/haskell/folds.scm:
```diff
@@ -0,0 +1,18 @@
+;; Copyright 2022 nvim-treesitter
```
Zed does not support folds unfortunately, this file should be removed.

mikayla-maki commented on crates/zed/src/languages/haskell/highlights.scm:
```diff
@@ -0,0 +1,632 @@
+(function
+  patterns:
+    (patterns
+      (_) @variable.parameter))
```
Unfortunately the Zed and neovim highlights files aren't compatible, as we use different conventions for syntax captures. You can find the list of captures that we support here: https://github.com/zed-industries/zed/blob/main/crates/theme/src/styles/syntax.rs

pseudomata replied:
This is a super helpful reference, thank you!

mikayla-maki commented on crates/zed/src/languages/haskell/injections.scm:
```diff
@@ -0,0 +1,97 @@
+  (quasiquote_body) @injection.content)
+
+((comment) @injection.content
+  (#set! injection.language "comment"))
```
I don't think we need this

mikayla-maki commented on crates/zed/src/languages/haskell/injections.scm:
```diff
@@ -0,0 +1,97 @@
+  (quoter) @_name
+  (#any-of? @_name "cassius" "lucius")
+  (quasiquote_body) @injection.content
+  (#set! injection.language "css"))
```
Zed uses `(#set! "language" "css")` instead of this syntax.

mikayla-maki commented on crates/zed/src/languages/haskell/injections.scm:
```diff
@@ -0,0 +1,97 @@
+  (quasiquote_body) @injection.content
+  (#set! injection.language "json"))
+
+; -----------------------------------------------------------------------------
```
Probably better to remove these comments wholesale :)

mikayla-maki reviewed (changes requested):
Thank you for adding Haskell support! Unfortunately, we can't merge this until this language fits with Zed's treesitter integrations. I think I've marked the main issues, the `highlights.scm` and `injections.scm` files. If the neovim highlights are giving you trouble, it might be worth trying the [tree-sitter-haskell highlights](https://github.com/tree-sitter/tree-sitter-haskell/blob/master/queries/highlights.scm) :)

Once you've got this to a good state, could you add a screenshot comparing syntax highlighting in your preferred editor and the same code's appearance in Zed? Thank you!

Commit:
Switch to `tree-sitter-haskell` grammar

pseudomata commented:
Thank you for the review @mikayla-maki. I've switched back to using the `highlights.scm` from `tree-sitter-haskell` as the base and modified it to use the [captures supported by Zed](https://github.com/zed-industries/zed/blob/main/crates/theme/src/styles/syntax.rs#L16-L68) (thank you for that reference!).

With these changes I've taken screenshots of a random code selection that showcases most of the syntax in common use, comparison is with the [Helix editor](https://helix-editor.com).

<details>
<summary>Screenshot showing syntax highlighting in Zed</summary>
<img src="https://github.com/zed-industries/zed/assets/132238190/3464c797-79c1-408c-b4cf-877846438c58">
</details>

<details>
<summary>Comparison screenshot showing syntax highlighting in Helix</summary>
<img src="https://github.com/zed-industries/zed/assets/132238190/be37d8bd-0d89-4928-b4d2-3bdacbd7af1a">
</details>

pseudomata commented on crates/zed/src/languages/haskell/injections.scm:
```diff
@@ -0,0 +1,62 @@
+; -----------------------------------------------------------------------------
+; General language injection
+(quasiquote
+  (quoter) @injection.language
+  (quasiquote_body) @injection.content)
```
It wasn't clear to me if there's different syntax for `@injection.language` and `@injection.content` to use here.

mikayla-maki replied:
Actually good point, I don't think we support this general injection query at all. It should be removed.

pseudomata commented on crates/zed/src/languages/haskell/injections.scm:
```diff
@@ -0,0 +1,62 @@
+;; Copyright 2022 nvim-treesitter
```
I wonder if it's worth it to just remove `injections.scm` from the PR for now since [`tree-sitter-haskell`](https://github.com/tree-sitter/tree-sitter-haskell) does not actually provide an `injections.scm` query file, so we're now mixing `tree-sitter-haskell` with `nvim-treesitter` 🤔

maxdeviant replied:
Yea, if `tree-sitter-haskell` doesn't support injections then we might as well remove `injections.scm` here.

mikayla-maki replied:
I think keeping the injections are OK, especially if these are well known.

maxdeviant commented on crates/zed/src/languages/haskell.rs:
```diff
@@ -0,0 +1,135 @@
+        if let Some(path) = last_binary_path {
+            Ok(LanguageServerBinary {
+                path,
+                arguments: Vec::new(),
```
I think we need to provide the same arguments to `LanguageServerBinary` that we do above.

We can do the same thing as what we do in the Gleam LSP adapter where we use a separate function to supply the necessary arguments: https://github.com/zed-industries/zed/blob/5db7e8f89ee4f5d9801a2eab88adc1db3fda0f46/crates/zed/src/languages/gleam.rs#L17-L19

pseudomata replied:
Done, thanks!

Commit:
Remove Haskell `injections.scm`

Since tree-sitter-haskell does not support injections we don’t want to mix with nvim-treesitter

Commit:
Use same arguments for LSP binary in `haskell.rs`

maxdeviant commented:
@pseudomata Have you had a chance to try out the LSP support in Zed?

@mikayla-maki and I are trying to test it out locally, but we're running into issues where it's not able to resolve the `haskell-language-server` for the right GHC version.

mikayla-maki reviewed (changes requested):
Like @maxdeviant said, we have been unable to test the LSP integration. I installed `hcup` via the [command on the documentation](https://www.haskell.org/ghcup/) and pointed Zed at the [PostgREST repo](https://github.com/PostgREST/postgrest). This PR would fail to find the other server binaries that the `haskell-language-server-wrapper` needs to work. We could get this to work by manually adding the Zed language server directory our `PATH`,  but this would immediately fail with a large error for multiple libraries:

```
server stderr: Some("dyld[61050]: Library not loaded: @rpath/libHSinteger-gmp-1.1-ghc9.4.8.dylib\n  Referenced from: <A63A67A5-EECB-3FDD-8198-240A3A4538FE> /Users/mikaylamaki/Library/Application Support/Zed/languages/haskell-language-server/haskell-language-server-2.6.0.0/bin/haskell-language-server-9.4.8\n
[snip]
```

What's worse is that these errors would not appear in the Zed log or the Zed language server logs, we could only find this by attempting to run `haskell-language-server-wrapper` from the command line. We also couldn't find a way to pass the binary folder path to the LSP via environment variables or configuration, but we might have missed something there.

As this has turned out to be both complicated and invisible to Zed UI, I recommend cutting the LSP part of this PR until we've built language extensions that y'all can use to solve these problems. But if you want to take a crack at fixing these problems, go right ahead :)

Thanks for being responsive and fixing things with us!

pseudomata commented:
Thank you @mikayla-maki and @maxdeviant. I believe this might have been the issue @leifu1128 was running into as well. I would like to take a crack at fixing the issues and see if I can come up with a reasonable solution first! I'm going to dig around a little bit and see if I can get things working by looking at how VS Code handles this (which might have a similar environment).

If not, I'll convert this PR to cover syntax highlighting only.

mikayla-maki commented:
Sounds good, and good luck!

pseudomata commented:
Aside: I just realized why LSP works fine for me -- it's because I have HLS installed via `ghcup` on my system, so `haskell-language-server-wrapper` is actually able to find the binary under `~/.ghcup/bin/`. In order to reproduce the issue, removing HLS via `ghcup rm hls 2.5.0.0` and re-running Zed will break things again.

mikayla-maki commented:
@pseudomata we could just request users install it themselves in our documentation. I believe for `nu` and `uiua`, whose LSPs are integrated into their binary, you just have to have it in your `PATH`. If Haskell land is ok with that, it's a good middle ground I think.

pseudomata commented:
I was actually just going to ask about this, I was looking at the Elixir implementation and copied that to expose a Haskell setting to do something similar, e.g. install HLS with `ghcup` and then adding this to editor configuration:
```json
  "haskell": {
    "lsp": {
      "local": {
        "path": "~/.ghcup/hls/2.5.0.0/bin/haskell-language-server-wrapper",
        "arguments": ["lsp"]
      }
    }
  }
```
To play it safe, I added the default to be `none` so a fresh installation of Zed would open a Haskell project with syntax highlighting but no LSP integration. So far things are working for me:

![CleanShot 2024-01-26 at 18 57 04@2x](https://github.com/zed-industries/zed/assets/132238190/64f39fd3-0ef9-4628-a2d8-c0ab62b743ef)

Let me clean up my branch really quick and I can push that change up.

Commit:
Ability to manually configure LSP for Haskell

Allow users who have HLS installed using `ghcup` to manually provide the path for the Haskell language server

pseudomata commented:
Let me know what you think! I see `nu` does this a bit differently (as opposed to exposing a setting), but considering we'd have to provide the path anyway which has the version in the path (`~/.ghcup/hls/2.5.0.0/bin/haskell-language-server-wrapper`) my thoughts are this avoid an error with the default setting and be more explicit:

```json
"haskell": {
  "lsp": "none"
}
```

But I'm happy to change it to match `nu` if that pattern is preferred.

mikayla-maki commented:
I think that the `nu` style with the useful error about `hls` would be best. No need for additional settings, smart default, works with the way most non-vscode haskell systems work. We can add more configuration options later if needed.

Commit:
Simplify Haskell LSP integration

This change will work without any configuration and assume the user will install the Haskell Language Server using `ghcup`

Commit:
Remove `xz` feature (no longer using it)

pseudomata commented:
Sounds good to me, thanks @mikayla-maki. I just pushed those changes up.

mikayla-maki commented on crates/zed/src/languages/haskell.rs:
```diff
@@ -0,0 +1,55 @@
+        _: &dyn LspAdapterDelegate,
+    ) -> Option<LanguageServerBinary> {
+        Some(LanguageServerBinary {
+            path: "haskell-language-server-wrapper".into(),
```
Is this still the correct binary to be invoking, now that we're sourcing it from the `PATH`?

pseudomata replied:
Yep, HLS ships with a few separate binaries for compatibility with different versions of GHC, the wrapper automatically selects the correct version.

mikayla-maki replied:
perfect, thank you!

mikayla-maki commented:
Oh and I'm so sorry, I forgot to mention this earlier, do you think you could add an `outline.scm` file? It's how a language integrates with Zed's buffer symbol search (`cmd-shift-o`).

pseudomata commented:
Oh cool, I hadn't discovered that feature yet! Happy to add that, I think it might take a bit to figure out all the symbols to add for Haskell (I just picked up learning the language so I'm a bit of a noob). I'm signing off for the weekend but might find some time to work on it over the next few days and can ping this PR again when I have something working -- I can also commit to doing that in a follow-up. Whichever you prefer 🙂

Thanks for the all the feedback, and for open sourcing Zed!

mikayla-maki commented:
Let's fast follow with it, y'all have put in the work for this one :). Thank you so much for the contribution!

leifu1128 commented:
Just woke up to see this, thanks for the hard work y'all. I can finally write more whitepapers.

BasLaa commented:
Is this already out in the main version of Zed?

maxdeviant commented:
> Is this already out in the main version of Zed?

Yes, it is available in Zed Stable starting from Zed 0.121.5.

maxdeviant commented:
It is currently bundled with part of Zed, but it will most likely move to an extension once we have language server support in extensions.

If it doesn't seem to be working then it might be a problem running the language server?

BasLaa commented:
It could be that, I'm not really sure of how I should be 'running' the language server if I'm honest. On VSCode it works well though, does that indicate anything?

pseudomata commented:
Hey @BasLaa, you need to have `HLS` installed (see [installation instructions via `ghcup`](https://haskell-language-server.readthedocs.io/en/latest/installation.html#ghcup)) and then things should work. If this isn't the case and you run into issues feel free to contact me and I'll see if I can help 🙂

BasLaa commented:
Hi @pseudomata, I have HLS installed via `ghcup`, although I do have multiple versions of HLS installed (2.4.0.0, 2.5.0.0, 2.6.0.0), that might interfere? Syntax highlighting for Haskell works for me in Zed, version is 0.123.6. I'm just trying out a Cabal project, and there I'm not getting any type signatures or hints, whereas I do get them in VSCode with the same project.

pseudomata commented:
I don't think multiple versions should be causing an issue, but I think it might be worth trying removing the others and re-installing just one version. AFAIK the 'wrapper' binary will select the correct version. 

Have you tried building the project first as well? It might also be helpful to see some of the debug logs and language server logs. Currently away from my computer so can't share exact steps to get logs but they're in the command panel.

BasLaa commented:
I tried removing all versions and reinstalling a single one, no change unfortunately. I also tried building the project. I believe I found the debug log for the LSP, but the log seems to be empty for me, so perhaps I'm looking at the wrong log?
